File: MainProject/Scenes/HexPlanet/Nodes/HexUnit.cs
using Apps.Queries.Abstractions.Tiles;
using Commons.Utils;
using Contexts;
using Domains.Models.Entities.Civs;
using Godot;
using GodotNodes.Abstractions.Addition;
using Infras.Readers.Abstractions.Nodes.Singletons;
using Infras.Writers.Abstractions.Civs;
using Infras.Writers.Abstractions.PlanetGenerates;
using Nodes.Abstractions;

namespace ZeromaXsPlaygroundProject.Scenes.HexPlanet.Nodes;

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH
/// Date: 2025-03-01 12:45
public partial class HexUnit : CsgBox3D, IHexUnit
{
    public HexUnit()
    {
        InitServices();
        var unit = _unitRepo!.Add();
        Id = unit.Id;
    }

    public NodeEvent? NodeEvent => null;

    #region 服务与存储

    private static ITileRepo? _tileRepo;
    private static ITileShaderApplication? _tileShaderApplication;
    private static IUnitRepo? _unitRepo;
    private static IHexPlanetManagerRepo? _hexPlanetManagerRepo;

    private static void InitServices()
    {
        _tileRepo ??= Context.GetBeanFromHolder<ITileRepo>();
        _tileShaderApplication ??= Context.GetBeanFromHolder<ITileShaderApplication>();
        _unitRepo ??= Context.GetBeanFromHolder<IUnitRepo>();
        _hexPlanetManagerRepo ??= Context.GetBeanFromHolder<IHexPlanetManagerRepo>();
    }

    #endregion

    public int Id { get; private set; }
    private Vector3 _beginRotation;
    private int _tileId;

    public int TileId
    {
        get => _tileId;
        set
        {
            if (_tileId > 0)
            {
                var preTile = _tileRepo!.GetById(_tileId)!;
                _tileShaderApplication!.DecreaseVisibility(preTile, Unit.VisionRange);
                _tileRepo.SetUnitId(preTile, 0);
            }

            _tileId = value;
            _unitRepo!.GetById(Id)!.TileId = _tileId;
            ValidateLocation();
            var tile = _tileRepo!.GetById(_tileId)!;
            _tileShaderApplication!.IncreaseVisibility(tile, Unit.VisionRange);
            _tileRepo.SetUnitId(tile, Id);
        }
    }

    private float _orientation;

    // 朝向（弧度制）
    public float Orientation
    {
        get => _orientation;
        set
        {
            _orientation = value;
            Rotation = _beginRotation;
            Rotate(Position.Normalized(), _orientation);
        }
    }

    private HexUnitPath? _path;
    private int _pathTileIdx;
    private bool _pathOriented;
    private const float PathRotationSpeed = Mathf.Pi;
    private const float PathMoveSpeed = 30f; // 每秒走 30f 标准距离

    public override void _Process(double delta)
    {
        if (_path == null) return;
        var deltaProgress = (float)delta * _hexPlanetManagerRepo!.StandardScale * PathMoveSpeed;
        if (_pathOriented)
        {
            var prePathTileIdx = _pathTileIdx;
            var progress = _path.GetProgress();
            while (_pathTileIdx < _path.Progresses.Count && _path.Progresses[_pathTileIdx] < progress)
                _pathTileIdx++;
            if (prePathTileIdx != _pathTileIdx)
            {
                _tileShaderApplication!.DecreaseVisibility(_path.Tiles![prePathTileIdx], Unit.VisionRange);
                _tileShaderApplication.IncreaseVisibility(_path.Tiles[_pathTileIdx], Unit.VisionRange);
            }

            var before = _path.Curve.SampleBaked(progress - deltaProgress, true);
            Node3dUtil.AlignYAxisToDirection(this, Position, alignForward: before.DirectionTo(Position));
        }
        else
        {
            var forward = Position.DirectionTo(_path.Curve.SampleBaked(deltaProgress, true));
            var angle = Math3dUtil.GetPlanarAngle(-Basis.Z, forward, Position, true);
            var deltaAngle = float.Sign(angle) * PathRotationSpeed * (float)delta;
            if (Mathf.Abs(deltaAngle) >= Mathf.Abs(angle))
            {
                Rotate(Position.Normalized(), angle);
                _pathOriented = true;
                _path.StartMove(this);
            }
            else
                Rotate(Position.Normalized(), deltaAngle);
        }
    }

    public void Travel(HexUnitPath path)
    {
        _path = path;
        _pathOriented = false;
        _pathTileIdx = 0;
        // 提前把实际单位数据设置到目标 Tile 中
        var fromTile = _tileRepo!.GetById(_tileId)!;
        _tileRepo.SetUnitId(fromTile, 0);
        var toTile = _path.Tiles![^1];
        _tileRepo.SetUnitId(toTile, Id);
        _unitRepo!.GetById(Id)!.TileId = toTile.Id;
        _tileId = toTile.Id;
    }

    public void FinishPath()
    {
        _path = null;
    }

    public void ValidateLocation()
    {
        var tile = _tileRepo!.GetById(_tileId)!;
        var position = tile.GetCentroid(_hexPlanetManagerRepo!.Radius + _hexPlanetManagerRepo.GetHeight(tile));
        Node3dUtil.PlaceOnSphere(this, position, _hexPlanetManagerRepo.StandardScale,
            alignForward: Vector3.Up); // 单位不需要抬高，场景里已经设置好了
        _beginRotation = Rotation;
    }

    public void Die()
    {
        _unitRepo!.Delete(Id);
        var tile = _tileRepo!.GetById(_tileId)!;
        _tileShaderApplication!.DecreaseVisibility(tile, Unit.VisionRange);
        _tileRepo.SetUnitId(tile, 0);
        QueueFree();
    }
}

File: TerraObserver/TO.Domains/Functions/Units/HexUnit.fs
namespace TO.Domains.Functions.Units

open System
open Friflo.Engine.ECS
open Godot
open TO.Domains.Functions.HexSpheres.Components
open TO.Domains.Functions.HexSpheres.Components.Tiles
open TO.Domains.Functions.Maths
open TO.Domains.Types.Configs
open TO.Domains.Types.Friflos
open TO.Domains.Types.HexSpheres
open TO.Domains.Types.HexSpheres.Components
open TO.Domains.Types.HexSpheres.Components.Tiles
open TO.Domains.Types.PathFindings
open TO.Domains.Types.Shaders
open TO.Domains.Types.Units

module HexUnit =
    let visionRange = 3

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-07-11 09:47:11
module HexUnitCommand =
    let private increaseVisibility
        (env: 'E when 'E :> ITileQuery and 'E :> ITileSearcherCommand and 'E :> ITileShaderDataCommand)
        (fromTileId: TileId)
        (range: int)
        =
        let fromTile = env.GetTile fromTileId
        let tileIds = env.GetVisibleTiles fromTile range

        for tileId in tileIds do
            let tile = env.GetTile tileId
            let tileVis = tile |> Tile.visibility
            let newTileVis = tileVis.Visibility + 1 |> TileVisibility
            tile.AddComponent(&newTileVis) |> ignore

            if newTileVis.Visibility = 1 then
                let newTileFlag =
                    tile |> Tile.flag |> TileFlag.withMask TileFlagEnum.Explored |> TileFlag

                tile.AddComponent(&newTileFlag) |> ignore

                env.RefreshTileShaderDataVisibility tile.Id
                <| Tile.countId tile
                <| Tile.flag tile
                <| Tile.visibility tile

    let private decreaseVisibility
        (env: 'E when 'E :> ITileQuery and 'E :> ITileSearcherCommand and 'E :> ITileShaderDataCommand)
        (fromTileId: TileId)
        (range: int)
        =
        let fromTile = env.GetTile fromTileId
        let tileIds = env.GetVisibleTiles fromTile range

        for tileId in tileIds do
            let tile = env.GetTile tileId
            let tileVis = tile |> Tile.visibility
            let newTileVis = tileVis.Visibility - 1 |> TileVisibility
            tile.AddComponent(&newTileVis) |> ignore

            if newTileVis.Visibility = 0 then
                env.RefreshTileShaderDataVisibility tile.Id
                <| Tile.countId tile
                <| Tile.flag tile
                <| Tile.visibility tile

    let pathRotationSpeed = Mathf.Pi
    let pathMoveSpeed = 30f // 每秒走 30f 标准距离

    let onHexUnitProcessed (env: #IPlanetConfigQuery) : OnHexUnitProcessed =
        fun (delta: float32) (unit: IHexUnit) ->
            if unit.Path <> null then
                let deltaProgress = delta * env.PlanetConfig.StandardScale * pathMoveSpeed

                if unit.PathOriented then
                    let prePathTileIdx = unit.PathTileIdx
                    let progress = unit.Path |> HexUnitPathQuery.getUnitPathProcess

                    while unit.PathTileIdx < unit.Path.Progresses.Length
                          && unit.Path.Progresses[unit.PathTileIdx] < progress do
                        unit.PathTileIdx <- unit.PathTileIdx + 1

                    if prePathTileIdx <> unit.PathTileIdx then
                        decreaseVisibility env unit.Path.TileIds[prePathTileIdx] HexUnit.visionRange
                        increaseVisibility env unit.Path.TileIds[unit.PathTileIdx] HexUnit.visionRange

                    let before = unit.Path.Curve.SampleBaked(progress - deltaProgress, true)

                    Node3dUtil.AlignYAxisToDirection(
                        unit,
                        unit.Position,
                        alignForward = before.DirectionTo unit.Position
                    )
                else
                    let forward =
                        unit.Position.DirectionTo <| unit.Path.Curve.SampleBaked(deltaProgress, true)

                    let angle = Math3dUtil.GetPlanarAngle(-unit.Basis.Z, forward, unit.Position, true)
                    let deltaAngle = float32 (Single.Sign angle) * pathRotationSpeed * delta

                    if Mathf.Abs deltaAngle >= Mathf.Abs angle then
                        unit.Rotate(unit.Position.Normalized(), angle)
                        unit.PathOriented <- true
                        unit.Path |> HexUnitPathCommand.unitStartMove unit
                    else
                        unit.Rotate(unit.Position.Normalized(), deltaAngle)

    let validateUnitLocation
        (env:
            'E
                when 'E :> IEntityStoreQuery
                and 'E :> ITileQuery
                and 'E :> IPlanetConfigQuery
                and 'E :> ICatlikeCodingNoiseQuery)
        : ValidateUnitLocation =
        fun (hexUnit: IHexUnit) ->
            let unitId = hexUnit.Id
            let unit = env.GetEntityById unitId
            let unitTileId = unit.GetComponent<UnitComponent>().TileId
            let tile = env.GetTile unitTileId

            let position =
                tile
                |> Tile.unitCentroid
                |> TileUnitCentroid.scaled (env.PlanetConfig.Radius + env.GetHeight tile)

            Node3dUtil.PlaceOnSphere(hexUnit, position, env.PlanetConfig.StandardScale, alignForward = Vector3.Up)
            hexUnit.BeginRotation <- hexUnit.Rotation

    let changeUnitTileId
        (env: 'E when 'E :> IEntityStoreQuery and 'E :> ITileQuery and 'E :> IHexUnitCommand)
        : ChangeUnitTileId =
        fun (tileId: TileId) (hexUnit: IHexUnit) ->
            let unitId = hexUnit.Id
            let unit = env.GetEntityById unitId
            let pre = unit.GetComponent<UnitComponent>().TileId

            if pre > 0 then
                decreaseVisibility env pre HexUnit.visionRange
                let emptyUnitId = TileUnitId 0
                (env.GetTile pre).AddComponent(&emptyUnitId) |> ignore

            let newUnitTileId = UnitComponent tileId
            unit.AddComponent(&newUnitTileId) |> ignore
            env.ValidateUnitLocation hexUnit
            increaseVisibility env tileId HexUnit.visionRange
            let newTile = env.GetTile tileId
            let newUnitId = TileUnitId unitId
            newTile.AddComponent(&newUnitId) |> ignore

    let killUnit (env: 'E when 'E :> ITileQuery and 'E :> IEntityStoreQuery) : KillUnit =
        fun (hexUnit: IHexUnit) ->
            let unit = env.GetEntityById hexUnit.Id
            let tileId = unit.GetComponent<UnitComponent>().TileId
            decreaseVisibility env tileId HexUnit.visionRange
            let tile = env.GetTile tileId
            let newTileUnitId = TileUnitId 0
            tile.AddComponent(&newTileUnitId) |> ignore
            unit.DeleteEntity()
            hexUnit.QueueFree()

    let travelUnit (env: 'E when 'E :> ITileQuery and 'E :> IEntityStoreQuery) : TravelUnit =
        fun (hexUnit: IHexUnit) (path: IHexUnitPath) ->
            hexUnit.Path <- path
            hexUnit.PathOriented <- false
            hexUnit.PathTileIdx <- 0
            // 提前把实际单位数据设置到目标 Tile 中
            let unit = env.GetEntityById hexUnit.Id
            let fromTile = env.GetTile <| unit.GetComponent<UnitComponent>().TileId
            let emptyTileUnitId = TileUnitId 0
            fromTile.AddComponent(&emptyTileUnitId) |> ignore
            let toTile = env.GetTile hexUnit.Path.TileIds[hexUnit.Path.TileIds.Length - 1]
            let newTileUnitId = TileUnitId hexUnit.Id
            toTile.AddComponent(&newTileUnitId) |> ignore
            let newUnitTileId = UnitComponent toTile.Id
            unit.AddComponent(&newUnitTileId) |> ignore
